### DISCUSSION TITLE
Component Model - support render function that captures state

### DISCUSSION THREAD

**Discussion by JaggerJo** [fsprojects/Avalonia.FuncUI#419] [Ideas]
Currently the render function of a component is only set once when the component is initially created. Even when the component is re-rendered the render function is not patched. This means that if the render function captures state that
state is never updated.

Take a look at the example code below:

<img width="905" alt="Screenshot 2024-04-14 at 14 54 26" src="https://github.com/fsprojects/Avalonia.FuncUI/assets/13090415/d59c1927-ffda-403a-a79f-35166399ca7c">

Running the code below with our current implementation of components produces a stale `counterView`. The `counterView` is never re-rendered.

With the changed applied from the linked PR the `counterView`'s render function is executed when the render function changes. 

As we don't want to patch the render function each time the component is re-rendered we need to detect if the render function captures state. I've included a prototype that does this via reflection per type and caches the result. I think we can make this work.

PR: https://github.com/fsprojects/Avalonia.FuncUI/pull/416

**Comment by Numpsy** [comment]
Sounds interesting, I'll try to have a look at the weekend
